services/src/main/java/org/keycloak/forms/login/freemarker/model/IdentityProviderBean.java

```diff
@@ -97,12 +128,121 @@ private String getLoginIconClasses(IdentityProviderModel identityProvider) {
+     * @param context a reference to the {@link AuthenticationFlowContext}.
+     * @return a {@link Set} containing the aliases of the IDPs that should be available for login. An empty set indicates
+     * that no IDPs should be available.
+     */
+    protected Set<String> getFederatedIdentities(KeycloakSession session, RealmModel realm, AuthenticationFlowContext context) {
+        Set<String> result = null;
+        if (context != null) {
+            UserModel currentUser = context.getUser();
+            if (currentUser != null) {
+                Set<String> federatedIdentities = session.users().getFederatedIdentitiesStream(session.getContext().getRealm(), currentUser)
+                        .map(FederatedIdentityModel::getIdentityProvider)
+                        .collect(Collectors.toSet());
+
+                if (!federatedIdentities.isEmpty() || organizationsDisabled(realm))
+                    // if orgs are enabled, we don't want to return an empty set - we want the organization IDPs to be shown if those are available.
+                    result = new HashSet<>(federatedIdentities);
+
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Builds and returns a list of {@link IdentityProvider} instances from the specified set of federated IDPs. The IDPs
+     * must be enabled, not link-only, and not set to be hidden on login page. If any IDP has an alias that matches the
+     * {@code existingIDP} parameter, it must be filtered out.
+     *
+     * @param federatedProviders a {@link Set} containing the aliases of the federated IDPs that should be considered for login.
+     * @param existingIDP the alias of the IDP that must be filtered out from the result (used when linking a new IDP to a user's account).
+     * @return a {@link List} containing the constructed {@link IdentityProvider}s.
+     */
+    protected List<IdentityProvider> getFederatedIdentityProviders(Set<String> federatedProviders, String existingIDP) {
+        return federatedProviders.stream()
+                .filter(alias -> !Objects.equals(existingIDP, alias))
+                .map(alias -> session.identityProviders().getByAlias(alias))
+                .filter(federatedProviderPredicate())
+                .map(idp -> createIdentityProvider(this.realm, this.baseURI, idp))
+                .sorted(IDP_COMPARATOR_INSTANCE).toList();
     }
 
-    public boolean isDisplayInfo() {
-        return realm.isRegistrationAllowed() || displaySocial;
+    /**
+     * Returns a predicate that can filter out IDPs associated with the current user's federated identities before those
+     * are converted into {@link IdentityProvider}s. Subclasses may use this as a way to further refine the IDPs that are
+     * to be returned.
+     *
+     * @return the custom {@link Predicate} used as a last filter before conversion into {@link IdentityProvider}
+     */
+    protected Predicate<IdentityProviderModel> federatedProviderPredicate() {
```

Comment:
Let's see if we can use the solution we discussed about having the login filters in a single place to make it easier to identify where brokers are filtered during the login as well as help when changing those filters.